uniffi_bindgen/src/bindings/kotlin/templates/ErrorTemplate.kt

```diff
@@ -1,149 +1,95 @@
+interface ErrorReader<E> {
+    fun read(buf: ByteBuffer): E;
 }
 
-class InternalException(message: String) : Exception(message)
-
 {%- for e in ci.iter_error_definitions() %}
-internal open class {{e.name()}} : RustError() {
-    class ByReference: {{e.name()}}(), RustErrorReference
 
-    @Suppress("ReturnCount", "TooGenericExceptionThrown", "UNCHECKED_CAST")
-    override fun<E: Exception> intoException(): E {
-        if (!isFailure()) {
-            // It's probably a bad idea to throw here! We're probably leaking something if this is
-            // ever hit! (But we shouldn't ever hit it?)
-            throw RuntimeException("[Bug] intoException called on non-failure!")
-        }
-        val message = this.consumeErrorMessage()
-        when (code) {
-            {% for value in e.values() -%}
-            {{loop.index}} -> return {{e|exception_name_kt}}.{{value}}(message) as E
-            {% endfor -%}
-            -1 -> return InternalException(message) as E
-            else -> throw RuntimeException("invalid error code passed across the FFI")
+// Error {{ e.name() }}
+{%- let toplevel_name=e.name()|exception_name_kt %}
+open class {{ toplevel_name }}: Exception() {
+    // Each variant is a nested class
+    {% for variant in e.variants() -%}
+    {% if !variant.has_fields() -%}
+    class {{ variant.name()|class_name_kt }} : {{ toplevel_name }}()
+    {% else %}
+    class {{ variant.name()|class_name_kt }}(
+        {% for field in variant.fields() -%}
+        val {{ field.name()|var_name_kt }}: {{ field.type_()|type_kt}}{% if loop.last %}{% else %}, {% endif %}
+        {% endfor -%}
+    ) : {{ toplevel_name }}()
+    {%- endif %}
+    {% endfor %}
+
+    companion object BufferReader : ErrorReader<{{ toplevel_name }}> {
+        override fun read(buf: ByteBuffer): {{ toplevel_name }} {
+            return when(buf.getInt()) {
+                {%- for variant in e.variants() %}
+                {{ loop.index }} -> {{ toplevel_name }}.{{ variant.name()|class_name_kt }}({% if variant.has_fields() %}
+                    {% for field in variant.fields() -%}
+                    {{ "buf"|read_kt(field.type_()) }}{% if loop.last %}{% else %},{% endif %}
+                    {% endfor -%}
+                {%- endif -%})
+                {%- endfor %}
+                else -> throw RuntimeException("invalid error enum value, something is very wrong!!")
+            }
         }
     }
 }
-
-open class {{e|exception_name_kt}}(message: String) : Exception(message) {
-    {% for value in e.values() -%}
-    class {{value}}(msg: String) : {{e|exception_name_kt}}(msg)
-    {% endfor %}
-}
-
 {% endfor %}
 
-// Helpers for calling Rust with errors:
+// Helpers for calling Rust
 // In practice we usually need to be synchronized to call this safely, so it doesn't
 // synchronize itself
-private inline fun <U, E: RustErrorReference> nullableRustCall(callback: (E) -> U?, err: E): U? {
```

Comment:
As far as I could tell, we never called `nullableRustCall` so I removed support for it.  Tell me if that's a bad idea.